blueocean-core-js/src/js/components/RunButton.jsx

```diff
@@ -0,0 +1,167 @@
+import { Icon } from 'react-material-icons-blue';
+
+import { RunApi as runApi } from '../';
+import { SseBus as sseBus } from '../';
+import { ToastService as toastService } from '../';
+import { buildRunDetailsUrl } from '../UrlBuilder';
+
+const stopProp = (event) => {
+    event.stopPropagation();
+};
+
+/**
+ * Run Buttons allows a pipeline or branch to be run and also be stopped thereafter.
+ */
+export class RunButton extends Component {
+
+    constructor(props) {
+        super(props);
+
+        this.subscriptionId = null;
+
+        this.state = {
+            running: false,
+            stopping: false,
+        };
+    }
+
+    componentDidMount() {
+        this.subscriptionId = sseBus.subscribeToJob(
+            (runData, event) => this._onJobEvent(runData, event),
+            (event) => this._filterJob(event)
+        );
+    }
+
+    componentWillReceiveProps(nextProps) {
+        this._updateState(nextProps);
+    }
+
+    componentWillUnmount() {
+        if (this.subscriptionId) {
+            sseBus.unsubscribe(this.subscriptionId);
+        }
+    }
+
+    _updateState(nextProps) {
+        const oldStatus = this.props.latestRun && this.props.latestRun.state || '';
+        const newStatus = nextProps.latestRun && nextProps.latestRun.state || '';
+
+        // if the state of the run changed, then assume it's no longer trying to stop
+        if (oldStatus !== newStatus) {
+            this.setState({
+                stopping: false,
+            });
+        }
+    }
+
+    _onJobEvent(runData, event) {
+        const name = decodeURIComponent(
+            event.job_ismultibranch ? event.blueocean_job_branch_name : event.blueocean_job_pipeline_name
+        );
+        const runId = event.jenkins_object_id;
+
+        if (event.jenkins_event === 'job_run_started') {
+            const runDetailsUrl = buildRunDetailsUrl(runData);
+
+            toastService.newToast({
+                text: `Started "${name}" #${runId}`,
+                action: 'Open',
+                onActionClick: () => {
+                    if (this.props.onNavigation) {
+                        this.props.onNavigation(runDetailsUrl);
+                    }
+                },
+            });
+        } else if (event.jenkins_event === 'job_run_ended' && runData.result === 'ABORTED') {
+            toastService.newToast({
+                text: `Stopped "${name}" #${runId}`,
+            });
+        }
+    }
+
+    _filterJob(event) {
+        return event.blueocean_job_rest_url === this.props.runnable._links.self.href;
+    }
+
+    _onRunClick() {
+        runApi.startRun(this.props.runnable);
+
+        const name = this.props.runnable.name;
+
+        toastService.newToast({
+            text: `Queued "${name}"`,
+        });
+    }
+
+    _onStopClick() {
+        if (this.state.stopping) {
+            return;
+        }
+
+        this.setState({
+            stopping: true,
+        });
+
+        runApi.stopRun(this.props.latestRun);
```

Comment:
NIT, maybe just `run`? Could this be used to stop something from the activity, which is a run?